driver-core/src/main/com/mongodb/internal/connection/IamAuthenticator.java

```diff
@@ -0,0 +1,275 @@
+    }
+
+    private static class IamSaslClient implements SaslClient {
+        private final MongoCredential credential;
+        private int step = -1;
+        private byte[] clientNonce = new byte[RANDOM_LENGTH];
+        private String httpResponse;
+
+        IamSaslClient(final MongoCredential credential) {
+            this.credential = credential;
+        }
+
+        @Override
+        public String getMechanismName() {
+            AuthenticationMechanism authMechanism = credential.getAuthenticationMechanism();
+            if (authMechanism == null) {
+                throw new IllegalArgumentException("Authentication mechanism cannot be null");
+            }
+            return authMechanism.getMechanismName();
+        }
+
+        @Override
+        public boolean hasInitialResponse() {
+            return true;
+        }
+
+        @Override
+        public byte[] evaluateChallenge(final byte[] challenge) throws SaslException {
+            step++;
+            if (step == 0) {
+                return computeClientFirstMessage();
+            }
+            if (step == 1) {
+                return computeClientFinalMessage(challenge);
+            } else {
+                throw new SaslException(format("Too many steps involved in the %s negotiation.", getMechanismName()));
+            }
+        }
+
+        @Override
+        public boolean isComplete() {
+            return step == 1;
+        }
+
+        @Override
+        public byte[] unwrap(final byte[] bytes, final int i, final int i1) throws SaslException {
+            throw new UnsupportedOperationException("Not implemented yet!");
+        }
+
+        @Override
+        public byte[] wrap(final byte[] bytes, final int i, final int i1) throws SaslException {
+            throw new UnsupportedOperationException("Not implemented yet!");
+        }
+
+        @Override
+        public Object getNegotiatedProperty(final String s) {
+            throw new UnsupportedOperationException("Not implemented yet!");
+        }
+
+        @Override
+        public void dispose() throws SaslException {
+            // nothing to do
+        }
+
+        private byte[] computeClientFirstMessage() throws SaslException {
+            new SecureRandom(this.clientNonce).nextBytes(this.clientNonce);
+
+            BsonDocument document = new BsonDocument()
+                    .append("r", new BsonBinary(this.clientNonce))
+                    .append("p", new BsonInt32('n'));
+            return toBson(document);
+        }
+
+        private byte[] computeClientFinalMessage(final byte[] serverFirst) throws SaslException {
+            final BsonDocument document = new RawBsonDocument(serverFirst);
+            final String host = document.getString("h").getValue();
+
+            final byte[] serverNonce = document.getBinary("s").getData();
+            if (serverNonce.length != (2 * RANDOM_LENGTH) || !Arrays.equals(Arrays.copyOf(serverNonce, RANDOM_LENGTH), this.clientNonce)) {
```

Comment:
Nit: Let's separate the two conditions into two if statements, with a unique message for each.  That way if there's a problem we'll know from the message which condition was violated.

Reply:
Done